### DISCUSSION TITLE
How to setup a mip package with multiple versions?

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#9831] [Libraries & Drivers] [answered]
You can now publish your own package to a github repo and write a package.json file to use with *mip* thanks to the great work of @jimmo! (see https://github.com/jimmo/micropython-mlx90640 for an example).
Can anyone point me in the right direction how I can handle multiple versions of a package?
Do I need to create a separate branch/tag?
What to do with the .json file when there are multiple versions?
Thanks!

**Comment by jimmo** [comment, marked as answer]
Thanks @saraverbeecke 

When you install a package.json-based package from github, the version specified will be interpreted as a git branch (or tag).

```py
>>> mip.install('github:jimmo/micropython-mlx90640', version='devel')
```

will fetch package.json from the devel branch (and then also implicitly use the devel branch for fetching referenced files).

So the simple answer is, tag your repo at each version, and then you're all set.

**Reply by saraverbeecke** [reply]
Ok thanks for the info! So I use a branch name or a commit tag and the package.json of that specific branch or tag will be downloaded. Is the "version" keyword in the package.json purely informative?
Do you know if this could work with Bitbucket as well?

**Reply by jimmo** [reply]
> So I use a branch name or a commit tag and the package.json of that specific branch or tag will be downloaded.

Yes. `"github:org/repo" version="foo"` will fetch raw.githubusercontent.com/org/repo/foo/package.json, which will give a set of urls, and the version=foo will then be subsequently used when expanding github: in those urls.

The version keyword in package.json is not currently used by `mip`.

> Do you know if this could work with Bitbucket as well?

Yes, if bitbucket has a similar scheme for raw.githubusercontent.com. See https://github.com/micropython/micropython-lib/blob/master/micropython/mip/mip.py#L60

**Reply by saraverbeecke** [reply]
Thanks again for your answer @jimmo . I don't think it will work with private repo's now. I'm willing to contribute some code but it seems that for Bitbucket or GitLab private repositories there is no support for a username:password https downloads anymore. We will have to use tokens instead, but this means extra arguments in the urequests module has to be used like injecting the right headers for the token... Any ideas how we should tackle this?

**Reply by riaanvddool** [reply]
> Thanks again for your answer @jimmo . I don't think it will work with private repo's now. I'm willing to contribute some code but it seems that for Bitbucket or GitLab private repositories there is no support for a username:password https downloads anymore. We will have to use tokens instead, but this means extra arguments in the urequests module has to be used like injecting the right headers for the token... Any ideas how we should tackle this?

This is how it is done in python pip: https://docs.readthedocs.io/en/stable/guides/private-python-packages.html#from-a-git-repository
   
    git+https://{token}@gitprovider.com/user/project.git@{version}
